lib/logstash/codecs/cef.rb

```diff
@@ -303,31 +220,47 @@ def handle(data, &block)
     message = unprocessed_data
 
     # Try and parse out the syslog header if there is one
-    if (cef_version = event.get('cefVersion')).include?(' ')
+    cef_version_field = @header_fields[0]
+    if (cef_version = event.get(cef_version_field)).include?(' ')
       split_cef_version = cef_version.rpartition(' ')
-      event.set('syslog', split_cef_version[0])
-      event.set('cefVersion', split_cef_version[2])
+      event.set(@syslog_header, split_cef_version[0])
+      event.set(cef_version_field, split_cef_version[2])
     end
 
     # Get rid of the CEF bit in the version
-    event.set('cefVersion', delete_cef_prefix(event.get('cefVersion')))
+    event.set(cef_version_field, delete_cef_prefix(event.get(cef_version_field)))
 
     # Use a scanning parser to capture the Extension Key/Value Pairs
     if message && message.include?('=')
       message = message.strip
 
       message.scan(EXTENSION_KEY_VALUE_SCANNER) do |extension_field_key, raw_extension_field_value|
         # expand abbreviated extension field keys
-        extension_field_key = MAPPINGS.fetch(extension_field_key, extension_field_key)
+        extension_field_key = @decode_mapping.fetch(extension_field_key, extension_field_key)
 
         # convert extension field name to strict legal field_reference, fixing field names with ambiguous array-like syntax
         extension_field_key = extension_field_key.sub(EXTENSION_KEY_ARRAY_CAPTURE, '[\1]\2') if extension_field_key.end_with?(']')
 
         # process legal extension field value escapes
         extension_field_value = raw_extension_field_value.gsub(EXTENSION_VALUE_ESCAPE_CAPTURE, '\1')
 
+        if extension_field_key == TIMESTAMP_FIELD
+          extension_field_value = normalize_timestamp(extension_field_value)
+        end
+
         event.set(extension_field_key, extension_field_value)
       end
+      # adapt @timestamp timezone to dtz
+      device_timezone_field = ecs_compatibility == :disabled ? "deviceTimeZone" : "[event][timezone]"
+      if event.include?(device_timezone_field) && ecs_compatibility != :disabled
+        t = event.get(TIMESTAMP_FIELD).time
+        #t is in UTC convert to local time to have the original hour/minute/second
```

Comment:
Here is an approach that uses Java 8's `java.time`, which appears to work well for a variety of use-cases. I plan to commit with tests in the morning.

1. capture all extension k/v to intermediate hash instead of setting directly on event
2. in ECS mode, intercept _any_ timestamp value and normalize it with captured `dtz` or value of `default_timezone` (a new parameter),
3. set remaining extension key/value on event 

Use-cases 
 - [x] `rt=Mar 17 2021 20:08:17.127 -08:00 dtz=America/New_York` -> trust parsed tz
 - [x] `rt=Mar 17 2021 20:08:17.127 dtz=America/New_York` -> use device timezone
 - [x] `rt=1614838408.517 dtz=America/New_York` -> use unix epoch

I'll be committing with tests in the morning.

~~~ ruby

  CEF_TIMESTAMP_FORMAT_PARSER = java.time.format.DateTimeFormatter.ofPattern("MMM dd[ yyyy] HH:mm:ss[.SSS][ zzz]")

  def normalize_timestamp(value, device_timezone_name)
    value = value.to_i if integer?(value)
    time = case value
           when Time then value
           when String
             cef_timestamp_formatter = CEF_TIMESTAMP_FORMAT_PARSER
             if device_timezone_name
               zone = java.time.ZoneId.of(device_timezone_name)
               cef_timestamp_formatter = cef_timestamp_formatter.with_zone(zone)
             end
             parsed_time = java.time.ZonedDateTime.parse(value, cef_timestamp_formatter).to_instant

             # Ruby's `Time::at(sec, microseconds_with_frac)`
             Time.at(parsed_time.get_epoch_second, Rational(parsed_time.get_nano, 1000))
           when Numeric then Time.at(Rational(value, 1000)) # in CEF time is in milliseconds since 1970
           else fail("Failed to normalize time `#{value.inspect}`")
           end

    LogStash::Timestamp.new(time)
  end
~~~

Reply:
Very good point! I like a lot your proposal to collect the data in the intermediate hash and reason about the presence of `dtz` and the format of the time field (if zoned or not). We should update also the doc once we in are in the final shape.